src/main/java/gregtech/api/util/CapesRegistry.java

```diff
@@ -0,0 +1,199 @@
+            if (capes == null) {
+                capes = new ArrayList<>();
+            }
+            capes.add(new ResourceLocation(capeLocation));
+            unlockedCapes.put(uuid, capes);
+        }
+
+        wornCapes.clear();
+        NBTTagList wornCapesTag = comp.getTagList("WornCapesValList", Constants.NBT.TAG_COMPOUND);
+        for (int i = 0; i < wornCapesTag.tagCount(); i++) {
+            NBTTagCompound tag = wornCapesTag.getCompoundTagAt(i);
+            String capeLocation = tag.getString("Cape");
+            if (capeLocation.isEmpty())
+                continue;
+            UUID uuid = tag.getUniqueId("UUID");
+            wornCapes.put(uuid, new ResourceLocation(capeLocation));
+        }
+    }
+
+    public static void initializeStorage(World world) {
+        MapStorage storage = world.getMapStorage();
+        instance = (CapesRegistry) storage.getOrLoadData(CapesRegistry.class, DATA_ID);
+
+        if (instance == null) {
+            instance = new CapesRegistry();
+
+            for(Map.Entry<UUID, List<ResourceLocation>> givenCapeSet : givenCapes.entrySet()) {
+                unlockedCapes.put(givenCapeSet.getKey(), givenCapeSet.getValue());
+            }
+
+            List<ResourceLocation> devReward = new ArrayList<>();
+            devReward.add(Textures.GREGTECH_CAPE_TEXTURE);
+
+
+            storage.setData(DATA_ID, instance);
+        }
+    }
+
+    public boolean isDirty() {
+        return true; // Doesn't work too well otherwise.
+    }
+
+    /**
+     * Allows one to check what capes a specific player has unlocked through CapesRegistry.
+     * @param uuid The player data used to get what capes the player has through internal maps.
+     * @return A list of ResourceLocations containing the cape textures that the player has unlocked.
+     */
+    public static List<ResourceLocation> getUnlockedCapes(UUID uuid) {
+        return unlockedCapes.get(uuid);
+    }
+
+
+    /**
+     * Links an advancement with a cape, which allows a player to unlock it when they receive the advancement.
+     * This should only be called on world load, since advancements are only accessible then.
+     * @param advancement A ResourceLocation pointing to the advancement that is to be used for getting a cape.
+     * @param cape The ResourceLocation that points to the cape that can be unlocked through the advancement.
+     * @param world The world that may contain the advancement used for getting a cape.
+     */
+    public static void registerCape(ResourceLocation advancement, ResourceLocation cape, World world) {
+        if (world instanceof WorldServer) {
+            try {
+                Field advManager = World.class.getDeclaredField("advancementManager");
+                advManager.setAccessible(true);
+                Advancement advObject = ((AdvancementManager) advManager.get(world)).getAdvancement(advancement);
+                capeAdvancements.put(advObject, cape);
```

Comment:
The dev and build environments are named differently because of the MCP.
This need to fix, `advancementManager` should be SRG name instead. (for build)

so, you'd better use the `AdvancementManager advancementManager = ObfuscationReflectionHelper.getPrivateValue(World.class, world, "field_srg_name");` instead of the getField your self. which can help you to get the private field with a srg_name for both dev and build environments.

`field_191951_C` or `field_191983_k`. I found two srg names of it, idk which one is the correct one, you need to test them.

Reply:
Besides, When you need frequent access a private field, the best way is to use AT file (resource/gregtech_at.cfg). Set it up, then rebuild the project, and it will make the corresponding fields public via ASM, which is much more efficient (compared to calling reflection per access). This is what Forge provides.

Reply:
`field_191951_C` worked for this. I don't think we really need an AT for this, given that this is only called during world load, and not for that many times.